### PR TITLE
49 - Basic voting functionality

### PR DESCRIPTION
**Pull request by ericrkuo**
https://github.com/ubclaunchpad/foodies/issues/49

New column `votes` added to `Promotion`

Two new endpoints
```
POST promotions/:id/upVote
   // SQL statement: Update promotion set votes = votes + 1 where id = :id
POST promotions/:id/downVote
   // SQL statement: Update promotion set votes = votes - 1 where id = :id
```

I did not implement any rate-limiting yet, I think this is something we can tackle separately once we see a higher number of users for our product.

### CODE REVIEW

**Comment by ericrkuo** [comment]
self reminder: add transactions after https://github.com/ubclaunchpad/foodies/pull/82 is merged

**Review by ld357 — APPROVED** [review]
Looks good! Are we planning to implement the getVotesPromotion in another PR?

**Comment by ericrkuo** [comment]
> Looks good! Are we planning to implement the getVotesPromotion in another PR?

Do you mean when we call `GET /promotions or GET /promotions/:id` we return the `votes` as well? Or a separate endpoint that just returns the field `votes` for a promotion.

File: backend/src/main/controller/PromotionController.ts [hunk 1]
```diff
@@ -164,4 +164,50 @@ export class PromotionController {
...
+    next: NextFunction
+  ): Promise<any> => {
+    try {
+      await getManager().transaction(async (transactionalEntityManager) => {
+        const id = await IdValidation.schema.validateAsync(request.params.id, {
+          abortEarly: false,
+        });
+        await transactionalEntityManager
+          .getCustomRepository(PromotionRepository)
+          .increment({ id: id }, 'votes', 1);
```
Comment (andrewzulaybar) [inline backend/src/main/controller/PromotionController.ts:183]:
```suggestion
          .increment({ id }, 'votes', 1);
```

Super minor improvement here but also not a big deal!

File: backend/src/main/controller/PromotionController.ts [hunk 2]
```diff
@@ -164,4 +164,50 @@ export class PromotionController {
...
+    next: NextFunction
+  ): Promise<any> => {
+    try {
+      await getManager().transaction(async (transactionalEntityManager) => {
+        const id = await IdValidation.schema.validateAsync(request.params.id, {
+          abortEarly: false,
+        });
+        await transactionalEntityManager
+          .getCustomRepository(PromotionRepository)
+          .decrement({ id: id }, 'votes', 1);
```
Comment (andrewzulaybar) [inline backend/src/main/controller/PromotionController.ts:206]:
```suggestion
          .decrement({ id }, 'votes', 1);
```
Same as above

File: backend/src/test/controller/PromotionController.test.ts [hunk 3]
```diff
@@ -273,6 +273,72 @@ describe('Unit tests for PromotionController', function () {
...
+
+    request(app)
+      .post(`/promotions/${promotion.id}/downVote`)
+      .expect(204)
+      .end((err) => {
+        if (err) return done(err);
+        return getManager().transaction(
+          'READ UNCOMMITTED',
+          async (transactionalEntityManager) => {
+            // check that promotion votes has incremented
```
Comment (andrewzulaybar) [inline backend/src/test/controller/PromotionController.test.ts:328]:
```suggestion
            // check that promotion votes has decremented
```
Another minor change!

**Review by andrewzulaybar — APPROVED** [review]
Just some minor suggestions and I also had a question: this implementation doesn't keep track of which user liked the promotion right? I'm thinking that this would be handled in a future improvement once we have login functionality and a way to determine which user is sending requests + prevent a user from liking it more than once.

Anyways, looks good to me and feel free to merge!

File: backend/src/test/controller/PromotionController.test.ts (hunk 3 above)
Comment (ericrkuo) [reply backend/src/test/controller/PromotionController.test.ts:328]: Oopsies thanks for catching this!

**Comment by ericrkuo** [comment]
> Just some minor suggestions and I also had a question: this implementation doesn't keep track of which user liked the promotion right? I'm thinking that this would be handled in a future improvement once we have login functionality and a way to determine which user is sending requests + prevent a user from liking it more than once.
> 
> Anyways, looks good to me and feel free to merge!

Yeah actually this is something we should discuss. Currently, we have the behaviour to save a promotion, so we have a mapping of which user saved which promotions. I'm thinking that can be separate from the concept of `votes` which is more so like a gauge for credibility (similar to Reddit posts).

image
So for example, the upvote/downvote of a Reddit post can be translated to the credibility of a promotion with votes. And saving a post would be equivalent to saving a promotion. 

Hmm, but now that I think about it, you're right in terms of how we don't prevent a user from voting more than once because we don't know whether a user upvoted/downvoted a promotion already... Maybe I can create a table that has a `userid` and `promotionId`, and it tells which user upvoted or downvoted a promotion

**Pull request merged**